zenoh/src/net/routing/interceptor/access_control.rs

```diff
@@ -0,0 +1,433 @@
+                return None;
+            }
+        } else if let NetworkBody::Request(Request {
+            payload: RequestBody::Query(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.get(key_expr) == Permission::Deny {
+                return None;
+            }
+        } else if let NetworkBody::Declare(Declare {
+            body: DeclareBody::DeclareSubscriber(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.declare_subscriber(key_expr) == Permission::Deny {
+                return None;
+            }
+        } else if let NetworkBody::Declare(Declare {
+            body: DeclareBody::DeclareQueryable(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.declare_queryable(key_expr) == Permission::Deny {
+                return None;
+            }
+        }
+        Some(ctx)
+    }
+}
+
+impl InterceptorTrait for EgressAclEnforcer {
+    fn compute_keyexpr_cache(&self, key_expr: &KeyExpr<'_>) -> Option<Box<dyn Any + Send + Sync>> {
+        Some(Box::new(key_expr.to_string()))
+    }
+    fn intercept(
+        &self,
+        ctx: RoutingContext<NetworkMessage>,
+        cache: Option<&Box<dyn Any + Send + Sync>>,
+    ) -> Option<RoutingContext<NetworkMessage>> {
+        let key_expr = cache
+            .and_then(|i| match i.downcast_ref::<String>() {
+                Some(e) => Some(e.as_str()),
+                None => {
+                    log::debug!("[ACCESS LOG]: Cache content was not of type String");
+                    None
+                }
+            })
+            .or_else(|| ctx.full_expr())?;
+
+        if let NetworkBody::Push(Push {
+            payload: PushBody::Put(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.put(key_expr) == Permission::Deny {
+                return None;
+            }
+        } else if let NetworkBody::Request(Request {
+            payload: RequestBody::Query(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.get(key_expr) == Permission::Deny {
+                return None;
+            }
+        } else if let NetworkBody::Declare(Declare {
+            body: DeclareBody::DeclareSubscriber(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.declare_subscriber(key_expr) == Permission::Deny {
+                return None;
+            }
+        } else if let NetworkBody::Declare(Declare {
+            body: DeclareBody::DeclareQueryable(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.declare_queryable(key_expr) == Permission::Deny {
+                return None;
+            }
+        }
+        Some(ctx)
+    }
+}
+pub trait AclActionMethods {
+    fn policy_enforcer(&self) -> Arc<PolicyEnforcer>;
+    fn interface_list(&self) -> Vec<i32>;
+    fn zid(&self) -> ZenohId;
+    fn flow(&self) -> Flow;
+
+    fn put(&self, key_expr: &str) -> Permission {
```

Comment:
Methods `put`/`get`/`declare_subscriber`/`declare_queryable` seems very similar.
Will be better to have some common method to real check implementation which will take key_expr and action.